src/cryo_challenge/_map_to_map/map_to_map_distance_matrix.py

```diff
@@ -12,22 +12,80 @@
-    maps_gt,
-    maps_submission,
-    map_to_map_distance,
-    chunk_size_gt=None,
-    chunk_size_submission=None,
-):
-    return torch.vmap(
-        lambda maps_gt: torch.vmap(
-            lambda maps_submission: map_to_map_distance(maps_gt, maps_submission),
-            chunk_size=chunk_size_submission,
-        )(maps_submission),
-        chunk_size=chunk_size_gt,
-    )(maps_gt)
-
-
+class MapToMapDistance:
+    def __init__(self, config):
+        self.config = config
+
+    def get_distance(self, map1, map2):
+        raise NotImplementedError()
+
+    def get_distance_matrix(self, maps1, maps2):
+        chunk_size_submission = self.config["analysis"]["chunk_size_submission"]
+        chunk_size_gt = self.config["analysis"]["chunk_size_gt"]
+        distance_matrix = torch.vmap(
+            lambda maps1: torch.vmap(
+                lambda maps2: self.get_distance(maps1, maps2),
+                chunk_size=chunk_size_submission,
+            )(maps2),
+            chunk_size=chunk_size_gt,
+        )(maps1)
+
+        return distance_matrix
+    
+    def get_computed_assets(self, maps1, maps2):
+        return {}
+
+class L2DistanceNorm(MapToMapDistance):
+    def __init__(self, config):
+        super().__init__(config)
+
+    def get_distance(self, map1, map2):
+        return torch.norm(map1 - map2)**2
+    
+class L2DistanceSum(MapToMapDistance):
+    def __init__(self, config):
+        super().__init__(config)
+
+    def get_distance(self, map1, map2):
+        return compute_cost_l2(map1, map2)
+    
+class Correlation(MapToMapDistance):
+    def __init__(self, config):
+        super().__init__(config)
+
+    def get_distance(self, map1, map2):
+        return compute_cost_corr(map1, map2) 
+
+class BioEM3dDistance(MapToMapDistance):
+    def __init__(self, config):
+        super().__init__(config)
+
+    def get_distance(self, map1, map2):
+        return compute_bioem3d_cost(map1, map2) 
+    
+class FSCDistance(MapToMapDistance):
+    def __init__(self, config):
+        super().__init__(config)
+        
+    def get_distance_matrix(self, maps1, maps2): # custom method
+        maps_gt_flat = maps1
+        maps_user_flat = maps2
+        n_pix = self.config["data"]["n_pix"]
+        maps_gt_flat_cube = torch.zeros(len(maps_gt_flat), n_pix**3)
+        mask = (
+            mrcfile.open(self.config["data"]["mask"]["volume"]).data.astype(bool).flatten()
+        )
+        maps_gt_flat_cube[:, mask] = maps_gt_flat
+        maps_user_flat_cube = torch.zeros(len(maps_user_flat), n_pix**3)
+        maps_user_flat_cube[:, mask] = maps_user_flat
+        
+        cost_matrix, fsc_matrix =  compute_cost_fsc_chunk(maps_gt_flat_cube, maps_user_flat_cube, n_pix)
+        self.stored_computed_assets = {'fsc_matrix': fsc_matrix}
+        return cost_matrix
+    
+    def get_computed_assets(self, maps1, maps2):
+        return self.stored_computed_assets # must run get_distance_matrix first
+    
 def run(config):
```

Comment:
This should be separate from the classes.

Reply:
I think the general distance classes (that external contributers will add) would have this template

class that takes in config, has methods `get_distance_matrix` and `get_computed_assets`.

You can read the FSCDistance class as an example of how someone would write get_distance_matrix and get_computed_assets so that they return what they should - even through they both come from one function (here inside get_distance_matrix, which just passes things to get_computed_assets through self. stored_computed_assets).

Reply:
Let me know if you see a better way.

I'm also looking into having get_distance_matrix take a global parameter.

This would be useful for a ResDistance that used the computed assets from the fsc distance (which haven't been written to disk yet, and so need to be passed through from memory somehow).

See this branch https://github.com/flatironinstitute/Cryo-EM-Heterogeneity-Challenge-1/blob/resolution_metric/src/cryo_challenge/_map_to_map/map_to_map_distance_matrix.py#L94

Reply:
That would be fine, just make to include the [@override](https://peps.python.org/pep-0698/) decorator.

Reply:
The override decorator was only introduced in 3.12. If I include it, then things will fail for 3.8-3.11.

I've removed it for now